### DISCUSSION TITLE
Question: Is GLSL -> GLSL (for opengl) ready?

### DISCUSSION THREAD

**Discussion by Nazariglez** [gfx-rs/naga#1839] [Q&A] [answered]
Hi everybody!

I have a project that uses `glsl_to_spriv` to compile shaders at compilation time (https://github.com/Nazariglez/notan/issues/88), and it's a pain for newcomers to have cmake installed sometimes. I was wondering if naga is ready to be used as a transpiler for a GLSL code to GLSL code, for OpenGL. My doubts comes from different things, it seems that most of the people is using naga with WPGU, using as input the WGPU shader lang, some project like veloren reverted the migration to naga after a few weeks, and most people seems to focus on GLSL for Vulkan and not OpenGL. 

I would love to implement a WGPU backend for my project, but I need to go step by step. The first step if possible will be move from spirv to naga using as input GLSL 450 and as output GLSL compatible with GL 3.3, ES 3, and WebGL 2. 

I apologize if this is a dumb question, I am not so much into shaders and shader compilation. Thanks!

**Comment by JCapucho** [comment, marked as answer]
This is an hard question to answer glsl is a big language and supporting all of it takes a lot of work but `naga` already supports a lot so it should support most of the inputs you throw at it (if not you can always open an issue ;).

Since the front ends convert everything to the same intermediate representation the glsl output should work fine whether you use wgsl or glsl as the input.

The versions you want to use are supported both on the frontend and backend.

Veloren did revert the `naga` changes but this was because of a bug in `spirv-cross`, the backend that `wgpu` previously used for shader translation, that caused msl output to be broken, this was caused because the shaders naga produced triggered a bad code path in it and since no one was interested in patching `spirv-cross` it was decided to revert it since `wgpu` would switch to `naga` in the future anyway (which it did).

Basically, it's a supported use case and should work but you need to test it and it might contain bugs.

**Comment by Gordon-F** [comment]
> as output GLSL compatible with GL 3.3, ES 3, and WebGL 2

`naga` used as shader translation for OpenGL (GL ES 3.0 and WebGL2) `wgpu` backend. You can see this in action, for example in [wgpu examples](https://github.com/gfx-rs/wgpu/tree/master/wgpu) or [bevy examples](https://github.com/bevyengine/bevy/tree/latest/examples#wasm). 

Please note, that OpenGL ES 3.0 backend available only for Linux & Android. On Windows it required angle lib (more info in [README](https://github.com/gfx-rs/wgpu#supported-platforms)).

**Comment by Nazariglez** [comment]
Got it, I may try to migrate to naga leaving spirv under a feature flag in case somebody wants to use spirv instead of naga for a while.

Do you think that using WGSL as input will be better? 

Thanks!

**Comment by Gordon-F** [comment]
> Do you think that using WGSL as input will be better?

WGSL is a great modern shader language, but spec still [under development](https://github.com/gpuweb/gpuweb/issues?q=is%3Aopen+is%3Aissue+label%3Awgsl). If it not scared you, you can try, otherwise I recommended to wait 1.0 version.

**Reply by Nazariglez** [reply]
Got it thank you so much! I may try it.

**Comment by Nazariglez** [comment]
We can close this, thank you so much for the kind answers.
